package bicpam.gui;

import javax.swing.JFrame;

public abstract class Interface extends JFrame {

	private static final long serialVersionUID = 1L;

    public String MatrixVar = open()+"The accepted file formats include attribute-relation (.ARFF) and standard matrices (such as .TXT). The first line of standard matrices should specify the column identifiers, while the first entry of each line should specify the row identifier. Tabular data can be either delimited by tabs, spaces or commas."+close();
    public String NetworkVar = open()+"BicPAMS accepts any input file format (such as .TXT or .SIG) assuming that: the first line specifies the column identifiers, and each other line specifies an interaction/entry within the network. An entry specifies the nodes and the association strength. Entries can be either delimited by tabs, spaces or commas. In addition to the file, the column index identifying the first node, second node and association strength needs to be inputted. Illustrating, for a network with header \"idProteinA,nameProteinA,idProteinB,nameProteinB,weight\", the user should fix (node1,node2,score) indexes as (0,2,4) or (1,3,4)."+close();
    public String OutputVar = open()+"An asynchronous webservice is invoked upon running BicPAMS. When the stopping criteria is met, a textual and graphical display of the outputs is provided. The user can select whether the output is displayed in the same HTML page (complete solution) or sent through e-mail (containing the textual display only up to a maximum size of 10Mb)."+close();
    public String AssumptionVar = open()+"The coherency assumption defines the correlation of values within a bicluster: constant, additive, multiplicative, order-preserving, plaid or symmetric. In constant models, an observed pattern (possibly containing different items) is preserved across rows (or columns). In additive or multiplicative models, shifting or scaling factors are allowed per row (or column) in order to allow meaningful variations of the original pattern. In order-preserving models, the values per row induce the same ordering across columns. A plaid model considers the cumulative effect of the contributions from multiple biclusters on areas where their rows and columns overlap. In symmetric models, symmetries of the observed patterns can be observed on rows (or columns)."+close();
    public String CohenrecyStrengthVar = open()+"The number of items determines the allowed deviations from expected values. Illustrating, a gene expression matrix parameterized with 5 items has 2 levels of activation ({1,2}), 2 levels of repression ({-1,-2}) and 1 level of unchanged expression ({0}). By going beyond the differential values, BicPAMS enables the discovery of non-trivial yet coherent and meaningful correlations. To maintain consistency, additive (multiplicative) models should be used with an uneven (even) number of items. When considering order-preserving models, the number of items should be increased to balance the degree of co-occurrences versus precedences."+close();
    public String QualityVar = open()+"This field specifies the maximum number of allowed noisy/missing elements (determining the minimum overlapping threshold for merging procedures). The tolerance of biclusters to noise can be additionally addressed using noise handlers (see mapping options) and alternative postprocessing procedures."+close();
    public String PatternRepresentationVar = open()+"Closed patterns (default option) enable the discovery of maximal biclusters (biclusters that cannot be extended without the need of removing rows and columns). The use of maximal patterns leads to biclusters with the columns' size maximized. This gives a preference towards flattened biclusters, possibly neglecting both vertical and smaller biclusters. Finally, the use of simple/all frequent patterns leads to biclustering solutions with a high number of biclusters (possibly contained by another bicluster), which can be useful to guide postprocessing steps. As the user specifies one of these three options, the available pattern miners are dynamically updated."+close();
    public String OrientationVar = open()+"Coherency can be either observed across rows (default) or columns (searches are applied on the transposed matrix). When the number of columns highly exceeds the number of rows (or vice-versa when searches are applied on the transposed matrix), pattern miners with vertical data formats such as Eclat should be preferred."+close();
    public String NormalizationVar = open()+"Depending on the properties of the input data, the user can either normalize data per Row, Column or for the Overall data elements or ignore normalization by selecting the None option. Both outliers and missing values are handled separately."+close();
    public String DiscretizationVar = open()+"Real-valued data needs to be discretized to apply pattern-based biclustering (see noise handling to understand how BicPAMs guarantees robustness to discretization drawbacks). The user can select the cut-off points of a Gaussian distribution (default) or fixed ranges of values (equal sized intervals after excluding outliers). Note that fixed ranges can lead to an imbalanced distribution of items. The user can bypass this option for symbolic data by selecting the None option."+close();
    public String NoiseHandlingVar = open()+"Multi-item assignments can be considered to handle deviations on the expected values within a bicluster caused by noise or discretization issues. By selecting this option, 2 items are assigned to elements with a value near a boundary of discretization (value in range c in[a,b] when min(b-c,c-a)/(b-a) is less than 25%). In this context, the data elements become associated with a varying number of items, thus increasing the size of data for analysis."+close();
    public String SymmetriesVar = open()+"This option should be selected if the input data is composed by positive and negative values (as it naturally affects the properties of the outputted biclusters). When using symmetric ranges, additive (multiplicative) models should be parameterized with an odd (even) number of items to guarantee consistent shifts (scales)."+close();
    public String MissingsVar = open()+"The user can specify what happens in the presence of missing values. Since BicPAMS is natively prepared to analyze sparse data, the Remove option (default) simply signals the algorithms to exclude missings from the searches. Alternatively, the Replace option uses WEKA's imputation methods to fill missings (the error of imputations can be minimized by simultaneously recurring to noise handlers). We suggest the use of Remove option for network data and other meaningfully sparse datasets since BicPAMS is able to discover biclusters with missing interactions (see Quality parameter)."+close();
    public String RemovalsVar = open()+"This option supports the possibility to remove uninformative elements within the inputted matrix (or uninformative interactions within the inputted network). Zero Entries can be selected to remove the {0}-items, while the Differential option is used to focus on items with high absolute value (e.g. {-3,-2,2,3} when |L|=6). Uninformative elements may correspond to: 1) weak interactions in networks, 2) unchanged expression, 3) healthy evaluations from clinical data, among others."+close();
    public String StoppingCriteriaVar = open()+"The search algorithm runs until any of the available stopping criteria is met. The available options are: 1) minimum number of biclusters before merging (default), 2) minimum covered area by the discovered biclusters (as a percentage of the elements of the input data matrix or network), and 3) minimum support threshold (minimum number of rows per bicluster specified as a fraction of overall rows). The value associated with the selected option should be additionally specified. We suggest the definition of a high number of biclusters (>50) as the default option, in order to guarantee an adequate exploration of the input dataset."+close();
    public String MinColumnsVar = open()+"The minimum number of columns per bicluster can be optionally inputted to promote efficiency and align the outputs according to user expectations. A good principle to fix this value is to use the square root of the number of columns (interactions per nodes) of the input matrix (network)."+close();
    public String NrIterationsVar = open()+"BicPAMS default behavior relies on a single iteration. For data with large coherent regions that may prevent the discovery of smaller (yet relevant) regions, the number of iterations can be optionally increased to guarantee their discovery. On every new iteration, 25% of the most selected data elements (from the biclusters discovered from the previous iteration) are removed to guarantee a focus on new regions. Three iterations already guarantee an adequate space exploration for hard data settings."+close();
    public String ScalabilityVar = open()+"The selection of scalability enhancements leads to the application of data partitioning techniques for very large data. Uninteresting elements can be complementary removed. Although searches are no longer exhaustive, optimality guarantees are provided (see publications)."+close();
    public String PatternMinerVar = open()+"The available pattern mining algorithms are dynamically provided based on the selected coherency assumption and pattern representation. Sequential pattern miners (SPM) are provided for order-preserving models: PrefixSpan and IndexSpan (an optimized algorithm able to explore efficiency gains from item-indexable properties) are made available for simple pattern representations, while BIDE+ is provided for closed pattern representations. Frequent itemset miners (FIM) are selected for the remaining coherency assumptions. AprioriTID, F2G (pattern-growth method for data with a large extent of coherent areas) and Eclat (vertical method for data with a high number of columns) are made available for simple pattern representations. CharmDiffsets, AprioriTID and CharmTID are made available for closed pattern representations, while CharmMFI with diffsets is provided for maximal pattern representations."+close();
    public String MergingVar = open()+"Different merging procedures are made available (according to Methods for the Discovery of Large Item-Indexable Sequential Patterns):  heuristic (default option) for an efficient and effective yet non-exhaustive merging; and combinatorial and multi-support FIM alternatives for an exhaustive yet more costly postprocessing step."+close();
    public String FilteringVar = open()+"Filtering is essential to guarantee compact solutions (applied after merging). A biclustered is filtered if it has not enough Dissimilar Elements, Dissimilar Rows or Dissimilar Columns against a larger bicluster. Considering a filtering option with 20% of dissimilar elements. In this context, biclusters sharing more than 80% of their elements against a larger bicluster are removed."+close();

	public static String initError = "Unfortunately we were not able to finish your request.\n";
	public static String finalError = "If the problem persists, please contact us with the inputted parameters and dataset (if not private) to: rmch AT tecnico.ulisboa.pt. We will get back to you as soon as possible.";
	public static String fileErrorMessage = initError+"File reading error: we were not able to process your request due to an error associated with the loading of the inputted dataset.\n"
			+"Description: exception while reading the input data file. Please guarantee that the format conforms to the accepted rules described in tutorial.";
	public static String preProcessingErrorMessage = initError+"Reason: Unexpected problem during the preprocessing of the inputted data. Please verify if the values of the uploaded data file are preventing the selected preprocessing options (e.g. normalization applied over rows/columns/nodes where all values are missing).\n"+finalError;
	public static String paramErrorMessage = "There was an unexpected problem with the processing of the inputted parameters.\n"+finalError;
	public static String zeroErrorMessage = "No biclusters were found in the dataset that satisfy the inputted parameters.\n"
			+"Solution #1: relax your stopping criteria.\nSolution #2: change the selected coherency and quality.\n"+finalError;
	public static String runRrrorMessage = initError+"Reason: Unexpected problem during the mining or postprocessing stages of BicPAMS.\n"
				+"This might be a problem caused by the current version of BicPAMS (e.g. non-supported combination of parameters).\n"+finalError
				+"Additional information: Your request was successfully interpreted and the dataset was successfully upload and preprocessed.";
	public static String outputErrorMessage = initError+"Display error: your query was computed successfully! However, there was an unexpected problem during the display of found biclusters.\n"+finalError;

	protected String aboutBicPAMS = "<div style=\"width:800px;\">BicPAMS is a powerful and flexible tool for the exploratory and unsupervised analysis of tabular and network data."
    		+"<br>BicPAMS integrates state-of-the-art contributions on pattern-based biclustering (including "
    		+"<a href=\"http://www.almob.org/content/9/1/27/abstract\">BicPAM</a>, <a href=\"http://web.ist.utl.pt/rmch/upload/publications/BicNET%20HenriquesMadeira15.pdf\">BicNET</a>, <a href=\"http://www.biomedcentral.com/content/pdf/1471-2105-15-130.pdf\">BicSPAM</a>, <a href=\"http://web.ist.utl.pt/rmch/upload/publications/ConstraintsBic%20HenriquesMadeira15.pdf\">BiC2PAM</a>, <a href=\"http://ieeexplore.ieee.org/xpl/articleDetails.jsp?tp=&arnumber=7001608\">BiP</a>, DeBi and BiModule)." 
    		+"<br>As such, BicPAMS is able to efficiently discover exhaustive and flexible structures of biclusters, with parameterizable coherency and robustness to noise and missings."
    		+"<br>BicPAMS has been successfully tested on biomedical data (inc. expression data, clinical data and biological networks) and social data. Former publications:"
    		+"<br><br><span style=\"font-size:12px\">"
    		+"<table><tr><td></td><td align=\"left\"><table><tr><td><a href=\"http://www.almob.org/content/9/1/27/abstract\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/pdf.gif\" alt=\"pdf\" /></a></td><td><a href=\"http://www.almob.org/content/9/1/27/citation\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/bibtex.gif\" alt=\"bibtex\" /></a></td><td><a href=\"xxx\"></a></td></tr></table></td><td></td><td align=\"left\">"
    		+"Rui Henriques and Sara C. Madeira, 2014, <a href=\"http://www.almob.org/content/9/1/27/abstract\">BicPAM: Pattern-based biclustering for biomedical data analysis</a>,<br> Algorithms for Molecular Biology, 9(1):27-, BioMed Central Ltd [10.1186/s13015-014-0027-z]</td><td></td></tr></table>"
    		+"<table><tr><td></td><td align=\"left\"><table><tr><td><a href=\"\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/pdf.gif\" alt=\"pdf\" /></a></td><td><a href=\"http://ieeexplore.ieee.org/xpl/articleDetails.jsp?tp=&arnumber=7001608\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/bibtex.gif\" alt=\"bibtex\" /></a></td><td><a href=\"http://ieeexplore.ieee.org/assets/v2/html/download_citations.html\"></a></td></tr></table></td><td></td><td align=\"left\">"
    		+"Rui Henriques and Sara C. Madeira, 2015, <a href=\"http://ieeexplore.ieee.org/xpl/articleDetails.jsp?tp=&arnumber=7001608\">Biclustering with Flexible Plaid Models to Unravel Interactions between Biological Processes</a>,<br> IEEE/ACM Transactions on Computational Biology and Bioinformatics, 12(4):738-752, IEEE/ACM[10.1109/TCBB.2014.2388206]</td><td></td></tr></table>"
    		+"<table><tr><td></td><td align=\"left\"><table><tr><td><a href=\"http://web.ist.utl.pt/rmch/upload/publications/BicNET%20HenriquesMadeira15.pdf\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/pdf.gif\" alt=\"pdf\" /></a></td><td><a href=\"http://link.springer.com/chapter/10.1007/978-3-662-48221-6_1#\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/bibtex.gif\" alt=\"bibtex\" /></a></td><td><a href=\"xxx\"></a></td></tr></table></td><td></td><td align=\"left\">"
    		+"Rui Henriques and Sara C. Madeira, 2015, <a href=\"http://web.ist.utl.pt/rmch/upload/publications/BicNET%20HenriquesMadeira15.pdf\">BicNET: Efficient Biclustering of Biological Networks to Unravel Non-Trivial Modules</a>,<br> In Algorithms in Bioinformatics, 1-15, LNCS Series, vol.9289, Springer Berlin Heidelberg [10.1007/978-3-662-48221-6_1]</td><td></td></tr></table>"
    		+"<table><tr><td></td><td align=\"left\"><table><tr><td><a href=\"http://web.ist.utl.pt/rmch/upload/publications/ConstraintsBic%20HenriquesMadeira15.pdf\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/pdf.gif\" alt=\"pdf\" /></a></td><td><a href=\"http://link.springer.com/export-citation/chapter/10.1007/978-3-319-23485-4_34\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/bibtex.gif\" alt=\"bibtex\" /></a></td><td><a href=\"xxx\"></a></td></tr></table></td><td></td><td align=\"left\">"
    		+"Rui Henriques and Sara C. Madeira, 2015, <a href=\"http://web.ist.utl.pt/rmch/upload/publications/ConstraintsBic%20HenriquesMadeira15.pdf\">Pattern-based Biclustering with Constraints for Gene Expression Data Analysis</a>,<br>In Comp. Methods in Bioinformatics and Systems Biology (EPIA-CMBSB), 326-339, LNAI Series, vol.9273, Springer [10.1007/978-3-319-23485-4_34]</td><td></td></tr></table>"
    		+"<table><tr><td></td><td align=\"left\"><table><tr><td><a href=\"http://www.biomedcentral.com/content/pdf/1471-2105-15-130.pdf\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/pdf.gif\" alt=\"pdf\" /></a></td><td><a href=\"http://www.biomedcentral.com/1471-2105/15/130/citation\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/bibtex.gif\" alt=\"bibtex\" /></a></td><td><a href=\"xxx\"></a></td></tr></table></td><td></td><td align=\"left\">"
    		+"Rui Henriques and Sara C. Madeira, 2014, <a href=\"http://www.biomedcentral.com/content/pdf/1471-2105-15-130.pdf\">BicSPAM: Flexible Biclustering using Sequential Patterns</a>,<br>BMC Bioinformatics, 15:130, BioMed Central Ltd [doi:10.1186/1471-2105-15-130]</td><td></td></tr></table>"
    		+"<table><tr><td></td><td align=\"left\"><table><tr><td><a href=\"http://www.sciencedirect.com/science/article/pii/S003132031500240X\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/pdf.gif\" alt=\"pdf\" /></a></td><td><a href=\"http://www.sciencedirect.com/science/article/pii/S003132031500240X\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/bibtex.gif\" alt=\"bibtex\" /></a></td><td><a href=\"x\"></a></td></tr></table></td><td></td><td align=\"left\">"
    		+"Rui Henriques, Claudia Antunes and Sara C. Madeira, 2015, <a href=\"http://www.sciencedirect.com/science/article/pii/S003132031500240X\">A Structured View on Pattern Mining-based Biclustering</a>,<br>Pattern Recognition, 48(12):3941-3958, Elsevier [http://dx.doi.org/10.1016/j.patcog.2015.06.018]</td><td></td></tr></table>"
    		+"<table><tr><td></td><td align=\"left\"><table><tr><td><a href=\"http://web.ist.utl.pt/rmch/upload/publications/F2G%20EfficientDiscoveryOfFullPatterns%20(Henriques2013).pdf\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/pdf.gif\" alt=\"pdf\" /></a></td><td><a href=\"http://www.ecmlpkdd2013.org/new-frontiers-in-mining-complex-patterns/\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/bibtex.gif\" alt=\"bibtex\" /></a></td><td><a href=\"xxx\"></a></td></tr></table></td><td></td><td align=\"left\">"
    		+"Rui Henriques, Sara C. Madeira and Claudia Antunes, 2013, <a href=\"http://web.ist.utl.pt/rmch/upload/publications/F2G%20EfficientDiscoveryOfFullPatterns%20(Henriques2013).pdf\">F2G: Efficient Discovery of Full-Patterns</a>,<br>In ECML/PKDD IW on New Frontiers to Mine Complex Patterns, Springer-Verlag, Prague, Czech Republic</td><td></td></tr></table>"
    		+"<table><tr><td></td><td align=\"left\"><table><tr><td><a href=\"http://web.ist.utl.pt/rmch/upload/publications/LargeItemIndexsableSP%20(HenriquesAntunesMadeira%20Jan2014).pdf\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/pdf.gif\" alt=\"pdf\" /></a></td><td><a href=\"http://link.springer.com/export-citation/chapter/10.1007/978-3-319-08407-7_7\"><img src=\"http://web.ist.utl.pt/~rmch/upload/icons/bibtex.gif\" alt=\"bibtex\" /></a></td><td><a href=\"xxx\"></a></td></tr></table></td><td></td><td align=\"left\">"
    		+"Rui Henriques, Claudia Antunes and Sara C. Madeira, 2014, <a href=\"http://web.ist.utl.pt/rmch/upload/publications/LargeItemIndexsableSP%20(HenriquesAntunesMadeira%20Jan2014).pdf\">Methods for the Efficient Discovery of Large Item-Indexable Sequential Patterns</a>,<br>Lecture Notes in Computer Science, 100-116, Springer International Publishing [10.1007/978-3-319-08407-7_7]</td></tr></table></span></div>";
    			
    protected String contactUs = "<div style=\"width:700px;\">The BicPAMS team appreciates your suggestions and the reporting of any problem that you may be able to detect."
    		+ "<br>For this purpose, clarifications or any other questions related with BicPAMS please email Rui Henriques: <b>rmch</b> [at] tecnico [dot] ulisboa [dot] pt"
    		+ "<br><br><b>BicPAMS Team:</b><br><a href=\"http://web.ist.utl.pt/rmch/\" target=\"_blank\">Rui Henriques</a> (responsible)<br>Francisco L. Ferreira (interface developer)<br><a href=\"http://web.ist.utl.pt/sara.madeira/\" target=\"_blank\">Sara C. Madeira</a> (supervision)"
    		+ "<br><br>For more information access the <a href=\"http://web.ist.utl.pt/rmch/bicpams/\" target=\"_blank\">BicPAMS webpage</a>.</div>";

    protected String testCases = "<div style=\"width:450px;\">For both testing purposes and to become familiar with the potentialities of BicPAMS, we make available the 10 testing scenarios on our <a href=\"http://web.ist.utl.pt/rmch/bicpams/\">BicPAMS webpage</a> (Case Tests tab). On this section from the webpage you can also download synthetic and real datasets to explore the BicPAMS software.<br>"
    		+"<br>We also suggest to take a look into the introductory <a href=\"https://youtu.be/QWFBwKHGz4o\">video tutorial</a>.</div>";

    protected String tutorialText = "<div style=\"width:700px;\">"
    		+"<br><b>Introductory <a href=\"https://youtu.be/QWFBwKHGz4o\">video tutorial</a></b> (with <a href=\"http://web.ist.utl.pt/rmch/software/bicpams/data/gyeast.arff\">gyeast.arff</a> dataset)<br>"
    		//+"<a href=\"https://youtu.be/QWFBwKHGz4o\">Video tutorial for the online interface</a> (with <a href=\"http://waw.inesc-id.pt:8080/bicpams/data/gyeast.arff\">gyeast.arff</a> dataset)<br>"
    		//+"<a href=\"https://youtu.be/QWFBwKHGz4o\">Video tutorial for the desktop application</a> (with <a href=\"http://waw.inesc-id.pt:8080/bicpams/data/gyeast.arff\">gyeast.arff</a> dataset)<br><br>"
    		//+ "Illustrative Example</b><table style=\"font-size: 12px;\">"
    		//+"<tr><td align=\"center\">Work in progress!</td></tr></table><br>"
    		+"<br><b>Input and Output</b><table style=\"font-size: 12px;\">" 
    		+"<tr><td align=\"center\"><em>Matrix</em></td><td>&nbsp;</td><td>"+MatrixVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Network</em></td><td>&nbsp;</td><td>"+NetworkVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Results</em></td><td>&nbsp;</td><td>"+OutputVar+"</td></tr>"
    		+"</table><br><b>Biclustering Properties</b><table style=\"font-size: 12px;\">" 
    		+"<tr><td align=\"center\"><em>Coherency<br>Assumption</em></td><td>&nbsp;</td><td>"+AssumptionVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Cohenrecy<br>Strength</em></td><td>&nbsp;</td><td>"+CohenrecyStrengthVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Quality<br>(Merging<br>Overlap)</em></td><td>&nbsp;</td><td>"+QualityVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Pattern<br>Representation</em></td><td>&nbsp;</td><td>"+PatternRepresentationVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Orientation</em></td><td>&nbsp;</td><td>"+OrientationVar+"</td></tr>"
    		+"</table><br><b>Pre-processing and Mapping Options</b><table style=\"font-size: 12px;\">" 
    		+"<tr><td align=\"center\"><em>Normalization</em></td><td>&nbsp;</td><td>"+NormalizationVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Discretization</em></td><td>&nbsp;</td><td>"+DiscretizationVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Noise<br>Handling</em></td><td>&nbsp;</td><td>"+NoiseHandlingVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Symmetries</em></td><td>&nbsp;</td><td>"+SymmetriesVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Missings<br>Handling</em></td><td>&nbsp;</td><td>"+MissingsVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Remove<br>Uninformative<br>Elements</em></td><td>&nbsp;</td><td>"+RemovalsVar+"</td></tr>"
    		+"</table><br><b>Mining Options</b><table style=\"font-size: 12px;\">" 
    		+"<tr><td align=\"center\"><em>Stopping<br>Criteria</em></td><td>&nbsp;</td><td>"+StoppingCriteriaVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Minimum<br>#Columns</em></td><td>&nbsp;</td><td>"+MinColumnsVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>#Iterations</em></td><td>&nbsp;</td><td>"+NrIterationsVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Pattern<br>Miner</em></td><td>&nbsp;</td><td>"+PatternMinerVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Scalability<br>Enhancer</em></td><td>&nbsp;</td><td>"+ScalabilityVar+"</td></tr>"
    		+"</table><br><b>Closing Options</b><table style=\"font-size: 12px;\">" 
    		+"<tr><td align=\"center\"><em>Merging</em></td><td>&nbsp;</td><td>"+MergingVar+"</td></tr>"
    		+"<tr><td align=\"center\"><em>Filtering</em></td><td>&nbsp;</td><td>"+FilteringVar+"</td></tr></table></div>";

	
	protected abstract String open();
	protected abstract String close();
}
